faas/loop.go

```diff
@@ -23,23 +23,56 @@ import (
 )
 
 func faasLoop(stream pb.FaasService_TriggerStreamClient, f HandlerProvider, errorCh chan error) {
+	var (
+		err     error
+		srvrMsg *pb.ServerMessage
+		ctx     *triggerContextImpl
+	)
+
+	// try our best to send an error back to the client if a handler panics.
+	defer func() {
+		if rErr := recover(); rErr != nil {
+			if rErr != nil {
+				fmt.Printf("an error was returned by the handler function: %v\n", rErr)
+
+				if ctx == nil || srvrMsg == nil {
+					fmt.Println("ctx and/or srvrMsg are nil")
+					errorCh <- fmt.Errorf("%v", rErr)
+					return
+				}
+
+				if ctx.http != nil {
+					ctx.http.Response.Body = []byte("Internal Server Error")
+					ctx.http.Response.Headers = map[string][]string{
+						"Content-Type": {"text/plain"},
+					}
+					ctx.http.Response.Status = 500
+				} else if ctx.event != nil {
+					ctx.event.Response.Success = false
+				}
+
+				err = faasSendResponse(stream, ctx, srvrMsg.GetId())
+				if err != nil {
+					fmt.Println("Error sending error response", err)
+				}
+
+				errorCh <- fmt.Errorf("%v", rErr)
+			}
+		}
+	}()
```

Comment:
The scope of this seems a bit off, we want to recover in the case where the panic occurs in the scope of handler so this should probably be scoped to the for loop so the stream may continue handling requests and the user application continues to run.

Reply:
mmm, I am not sure that handling panics and continuing is a good idea..

Reply:
OK, much improved now :-)